### PR TITLE
feat: [SFEQS-1542] Add @pagopa/logger

### PR DESCRIPTION
**Pull request by lucacavallaro**
#### List of Changes

1. Add the new package `@pagopa/logger`

#### Motivation and Context

We needed a logging utility for our `fp-ts`-based projects

#### How Has This Been Tested?

The packages was fully unit-tested (see `src/__test__`)

#### Checklist:

- [ ] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.

### CODE REVIEW

**Comment by balanza** [comment]
I see we made something similar in https://github.com/pagopa/winston-ts (cc @fabriziopapi).

The big difference is that's opinionated on `winston` (for historical reasons I guess). Nevertheless, I like the API as I find it more _brainless_ to use.

Maybe we can have the best of both worlds? Maybe we can have the easy API of `winston-ts` without the dependency on `winston`?

File: packages/logger/package.json [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+    "build": "tsup-node",
+    "format": "prettier --write .",
+    "test": "vitest run --coverage",
+    "lint": "eslint \"src/**\"",
+    "lint:fix": "eslint --fix \"src/**\"",
+    "typecheck": "tsc"
+  },
+  "dependencies": {
+    "fp-ts": "^2.13.1",
+    "tsup": "^6.6.3"
```
Comment (balanza) [inline packages/logger/package.json]: I'm not sure is a production dependency.

Comment (lucacavallaro) [reply packages/logger/package.json]: thanks

**Comment by lucacavallaro** [comment]
> I see we made something similar in https://github.com/pagopa/winston-ts (cc @fabriziopapi).
> 
> The big difference is that's opinionated on `winston` (for historical reasons I guess). Nevertheless, I like the API as I find it more _brainless_ to use.
> 
> Maybe we can have the best of both worlds? Maybe we can have the easy API of `winston-ts` without the dependency on `winston`?

Imho these libraries serve a different purpose! this is an hexagonal "port" that helps developers to use their favorite logger in a `pure` and `stateless` way, `winston-ts` is a concrete implementation of a logger.

The `winston-ts` API and DX can be obtained easily componing this package's `log` function with the fp-ts utilities that makes the `lift` from IO and `chain`, and providing a `Logger` instance based on `winston` logger. Maybe we can release in the future this adapter as `@pagopa/logger-winston-ts` to help `winston-ts` consumer with the migration.

File: packages/logger/src/logger.ts [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+export type Logger = {
+  log: (s: string, l: LogRecord["level"]) => () => void;
+  format: (r: LogRecord) => string;
+  context?: Record<string, unknown>;
+};
+
+export const log =
+  (level: LogRecord["level"]) =>
+  (message: LogRecord["message"], context: Record<string, unknown> = {}) =>
+  (r: { logger: Logger }) =>
```
Comment (grausof) [inline packages/logger/src/logger.ts:20]: I would add a comment on this part. What is `r`?

**Review by grausof — DISMISSED** [review]
LGTM

File: packages/logger/src/logger.ts (hunk 2 above)
Comment (lucacavallaro) [reply packages/logger/src/logger.ts:20]: `r` is the dependencies object, `r` as `Reader`

Comment (grausof) [reply packages/logger/src/logger.ts:20]: I think it's necessary to always log the error like debugging the request.

**Review by grausof — APPROVED** [review]

**Comment by balanza** [comment]
The point of my comment is to log things like this:
```ts
pipe(
  RTE.right(name),
  logger.rte.info(name => `greet called ${name}`)
  RTE.map(name => `Hello ${name}`),
);
```

Instead of this
```ts
pipe(
  RTE.right(name),
  RTE.chainFirst((name) => L.infoRTE(`greet called ${name}`))
  RTE.map(name => `Hello ${name}`),
);
```

Don't you see any value in this API?

**Comment by lucacavallaro** [comment]
> The point of my comment is to log things like this:
> 
> ```ts
> pipe(
>   RTE.right(name),
>   logger.rte.info(name => `greet called ${name}`)
>   RTE.map(name => `Hello ${name}`),
> );
> ```
> 
> Instead of this
> 
> ```ts
> pipe(
>   RTE.right(name),
>   RTE.chainFirst((name) => L.infoRTE(`greet called ${name}`))
>   RTE.map(name => `Hello ${name}`),
> );
> ```
> 
> Don't you see any value in this API?

I find it too "implicit" it's not clear what it exactly does, but this kind of API can be obtained extracting the step in the chaining, like:

```typescript
const info = (msg: string) => RTE.chainFirst((name) => L.infoRTE(msg))
const rte = { info };
```

in my opinion it's out of the scope of this package, the abstractions to use for logging (and how to handle the concrete implementation) are up to the project

this package should be seen as a "skeleton" on which to build your logger

**Comment by balanza** [comment]
>  it's not clear what it exactly does

It logs a message with the appropriate level, and that's everything you want to know at the application abstraction level 🤗

Isn't the purpose of a package, to hide implementation choices from behavior choices? (like, for example, to use `chainFirst` instead of `chain` or `map`).

**Comment by lucacavallaro** [comment]
> Isn't the purpose of a package, to hide implementation choices from behavior choices? (like, for example, to use chainFirst instead of chain or map).

nope, the purpose of this package is to give a `pure` logger that can be used in `fp-ts` based projects

logging is an effectful action (wrapped in `IO<void>`) that should be treated in the same way we use `http clients`

dictating how it is composed within your application is not the task of a logger, it is something that must be managed at the application level by creating the necessary abstractions if necessary

**Comment by balanza** [comment]
In the example, why did you choose `chainFirst` instead of `chain` or `map`? You surely did that on purpose...

Isn't this knowledge worth being _standardized_? Or do we need such nuances when implementing business rules?

**Comment by gunzip** [comment]
If I undestand well, what @lucacavallaro is trying to say is that of course you can standardize the 
behavior using some `tap` utility, but that's not the _purpose_ of the logger module. 

Let's say you have something like this:

```
pipe(
  doSomething,
  tap(log),
)

```
the logger module is supposed to implement the `log` part, not the `tap` one.

As a side note, here is an in depth article about interleaving logging: 

https://dev.to/mstn/tap-ts-type-safe-eavesdropping-2jfp

we probably want to consider the async effectful side as well.

**Pull request merged**